File: hscripts/virtualbox-import-all.pl
#!/usr/bin/env perl
use strict;
use warnings;

use File::Which;
use File::Basename qw(dirname);
use feature 'say';

my $script_dir = dirname(__FILE__);
my $config = do "$script_dir/../data/setup-private.pl" or die "Failed to load setup-private.pl: $!";
my $virtualbox_dir = $config->{_private_virtualbox_dir};

if (defined which('VBoxManage')) {
	`VBoxManage setproperty machinefolder "$virtualbox_dir"`;
} else {
	die "Must have command 'VBoxManage' installed";
}

if (grep(/^--help|-h$/, @ARGV)) {
	say 'Usage: ' . $0 . ' [-h|--help] [--unregister]';
	exit 0;
}

if (grep(/^--unregister$/, @ARGV)) {
	open(my $fh, '-|', 'VBoxManage list vms') or die $!;
	while (my $line = <$fh>) {
		if ($line =~ m/\"(.*?)\" \{(.*?)\}/) {
			my $name = $1;
			my $uuid = $2;
			say('Removing "' . $name . '"');
			`VBoxManage unregistervm $uuid`
		} else {
			die "Capture group failed on line: $line";
			exit 0;
		}
	}
	exit 0;
}

if (! -d $virtualbox_dir) {
	die 'Failed to find directory: ' . $virtualbox_dir;
}

opendir(my $dh, $virtualbox_dir) or die $!;
while (my $dirname = readdir($dh)) {
	next if ($dirname eq '.' || $dirname eq '..');

	# Is not a group.
	if (-f "$virtualbox_dir/$dirname/$dirname.vbox") {
		say "REGISTERING $virtualbox_dir/$dirname/$dirname.vbox";
		`VBoxManage registervm "$virtualbox_dir/$dirname/$dirname.vbox"`;
	} else {
		opendir(my $dh2, "$virtualbox_dir/$dirname") or dir $!;
		while (my $dirname2 = readdir($dh2)) {
			next if ($dirname2 eq '.' || $dirname2 eq '..');

			if (-f "$virtualbox_dir/$dirname/$dirname2/$dirname2.vbox") {
				say "REGISTERING $virtualbox_dir/$dirname/$dirname2/$dirname2.vbox";
				`VBoxManage registervm "$virtualbox_dir/$dirname/$dirname2/$dirname2.vbox"`;
			}
		}
	}
}
closedir($dh);

File: scripts/virtualbox-import-all.py
#!/usr/bin/env python3
import argparse
import re
import shutil
import subprocess
import sys
from pathlib import Path

parser = argparse.ArgumentParser(prog='virtualbox-import-all')
parser.add_argument('--unregister', action='store_true')
args = parser.parse_args()

if shutil.which('VBoxManage') is None:
	print("Must have command 'VBoxManage' installed", file=sys.stderr)
	sys.exit(1)

config_path = Path(__file__).resolve().parent.parent / 'config' / 'setup-private.pl'
if not config_path.is_file():
	print(f'Failed to load setup-private.pl: {config_path}', file=sys.stderr)
	sys.exit(1)

match = re.search(r"_private_virtualbox_dir\s*=>\s*'([^']+)'", config_path.read_text())
if not match:
	print('Failed to find _private_virtualbox_dir in setup-private.pl', file=sys.stderr)
	sys.exit(1)

virtualbox_dir = Path(match.group(1))
subprocess.run(['VBoxManage', 'setproperty', 'machinefolder', str(virtualbox_dir)], check=False)

if args.unregister:
	proc = subprocess.run(
		['VBoxManage', 'list', 'vms'],
		capture_output=True,
		text=True,
		check=True,
	)
	for line in proc.stdout.splitlines():
		vm_match = re.match(r'^"(.+)" \{(.+)\}$', line)
		if not vm_match:
			print(f'Capture group failed on line: {line}', file=sys.stderr)
			sys.exit(1)
		name, uuid = vm_match.groups()
		print(f'Removing "{name}"')
		subprocess.run(['VBoxManage', 'unregistervm', uuid], check=False)
	sys.exit(0)

if not virtualbox_dir.is_dir():
	print(f'Failed to find directory: {virtualbox_dir}', file=sys.stderr)
	sys.exit(1)

for entry in virtualbox_dir.iterdir():
	if not entry.is_dir():
		continue

	vbox_file = entry / f'{entry.name}.vbox'
	if vbox_file.is_file():
		print(f'REGISTERING {vbox_file}')
		subprocess.run(['VBoxManage', 'registervm', str(vbox_file)], check=False)
		continue

	for subentry in entry.iterdir():
		if not subentry.is_dir():
			continue

		sub_vbox = subentry / f'{subentry.name}.vbox'
		if sub_vbox.is_file():
			print(f'REGISTERING {sub_vbox}')
			subprocess.run(['VBoxManage', 'registervm', str(sub_vbox)], check=False)
